### PR TITLE
Feature - Schema directives visitor

### PR DESCRIPTION
**Pull request by 0x1EE7**
@rafalp Here is a the first draft for porting the SchemaDirectiveVisitor from apollo into ariadne as we have discussed in #167 and related to #138 

While the simple test cases provided with this PR is now passing there is some TODO items that needs to be discussed and resolved. I will add some more test cases and also looking forward to your feedback.

### TODOs
- [x] Check if `update_each_key` can only accept a dict. `graphql-core` seems to be always returning dicts
- [x]  Needs improvements to remove nodes in `call_method`. Since python does not have 'undefined', False value is chosen for removal. Still more clarification is needed. Maybe return a tuple instead? But that would have an effect on the visitor methods' signature too.
- [x] How would `heal_schema` be needed?

### CODE REVIEW

File: ariadne/directives.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class DirectiveType(SchemaBindable):
+    def __init__(self, name: str, visitor: Type[SchemaDirectiveVisitor]) -> None:
+        self.name = name
+        self.visitor = visitor
+
+    def bind_to_schema(self, schema: GraphQLSchema) -> None:
+        SchemaDirectiveVisitor.visit_schema_directives(
+            schema, {self.name: self.visitor}
```
Comment (rafalp) [inline ariadne/directives.py]: Could we add `directives: Dict[Str, Type[SchemaDirectiveVisitor]] = None` argument to the `make_executable_schema` instead? That would give us the safety of running directives only after all other bindables set their resolvers/etc/ect.

File: tests/test_directives.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+        def resolve_reverse(obj, info, **kwargs):
+            result = original_resolver(obj, info, **kwargs)
+            return result[::-1]
+
+        field.resolve = resolve_reverse
+        return field
+
+
+def test_single_directive_without_args():
```
Comment (rafalp) [inline tests/test_directives.py]: Change test names to have the name of tested logic and result state: `test_field_definition_directive_replaces_field_resolver_with_custom_one`.

**Comment by rafalp** [comment]
> Needs improvements to remove nodes in call_method. Since python does not have 'undefined', False value is chosen for removal.

`False` is good replacement IMHO. Alternatively, we could pack than into a more descriptive constant (`REMOVED_TYPE = False`) or create an empty type like `class RemovedType` and test for instance/type equality, and check for that instead., but thats also going too magical for me.

> How would heal_schema be needed?

From my understanding Apollo's `healSchema` iterates schema types (and their fields) to make sure that following remains true after directives have been run:

```
GraphQLSchema({"User": GraphQLNamedType("User"))
```

Because directives can change schema structure, they can produce broken schema like this:

```
GraphQLSchema({"User": GraphQLNamedType("Author"))
```

So yes, we will need to eventually implement `heal_schema` too.

**Comment by 0x1EE7** [comment]
That's great feedback @rafalp I will look into `heal_schema` too.

File: ariadne/schema_visitor.py [hunk 3]
```diff
@@ -149,7 +151,7 @@ def visit_schema(
     ],
 ) -> GraphQLSchema:
     """
-    Helper function that calls visitor_selector and applies the resulting
+    Helper def that calls visitor_selector and applies the resulting
```
Comment (rafalp) [inline ariadne/schema_visitor.py]: "function" was better here ;)

Comment (0x1EE7) [reply ariadne/schema_visitor.py]: Haha definitely :)

File: ariadne/executable_schema.py [hunk 4]
```diff
@@ -1,21 +1,23 @@
...
 
 from .enums import set_default_enum_values_on_schema
+from .schema_visitor import SchemaDirectiveVisitor
 from .types import SchemaBindable
 
 
 def make_executable_schema(
     type_defs: Union[str, List[str]],
     bindables: Union[SchemaBindable, List[SchemaBindable], None] = None,
+    directives: Dict[str, Type[SchemaDirectiveVisitor]] = None,
```
Comment (rafalp) [inline ariadne/executable_schema.py]: Let's make directives option explicit:

```suggestion
    *,
    directives: Dict[str, Type[SchemaDirectiveVisitor]] = None,
```

**Comment by 0x1EE7** [comment]
@rafalp Is there anything more I could do to move this forward?

**Comment by rafalp** [comment]
The plan currently is to release Ariadne 0.6 as soon as possible and then focus releasing custom directives work from this PR as Ariadne 0.7.

**Comment by rafalp** [comment]
Ariadne 0.6 is out, so I'll clone this PR locally sometime next week and test it out myself to see how it works. 🥽

**Comment by 0x1EE7** [comment]
> Ariadne 0.6 is out, so I'll clone this PR locally sometime next week and test it out myself to see how it works. 🥽

Great, looking forward to it! 🙇

**Comment by rafalp** [comment]
I've did some quick experiments with this and it works great, but I've noticed you have pushed bunch of fixes yesterday. How do you feel about readyness of this for merging?

This feature is gigantozauric, so its gonna be fun to write reasonable tests harness and docs for it 😅I would like to merge it and then follow up with improvements and tests in next PRs. Perhaps bump version number to 0.7 and release few alphas in the meantime so others can also play with it.

**Comment by 0x1EE7** [comment]
> I've did some quick experiments with this and it works great, but I've noticed you have pushed bunch of fixes yesterday. How do you feel about readyness of this for merging?
> 
> This feature is gigantozauric, so its gonna be fun to write reasonable tests harness and docs for it 😅I would like to merge it and then follow up with improvements and tests in next PRs. Perhaps bump version number to 0.7 and release few alphas in the meantime so others can also play with it.

I think that's a great plan. The changes I pushed yesterday was very minimal in order to fix some failures that I noticed while developing some directives on top of this PR. I think it is ready from that point of view and more PRs could follow after this one is merged. 🌟

**Comment by rafalp** [comment]
I've released Ariadne `0.7.0a1`with this PR merged so if anyone wants to give this a try, they can install `ariadne==0.7.0a1`

**Pull request merged**